carma_cooperative_perception/src/multiple_object_tracker_component.cpp

```diff
@@ -452,6 +453,92 @@ static auto predict_track_states(std::vector<Track> tracks, units::time::second_
   return tracks;
 }
 
+struct SemanticDistance2dScore
+{
+  template <typename Track, typename Detection>
+  auto operator()(const Track & track, const Detection & detection) const -> std::optional<float>
+  {
+    if constexpr (std::is_same_v<decltype(track.state), decltype(detection.state)>) {
+      const auto dist{two_dimensional_distance(track.state, detection.state)};
+      if (
+        track.semantic_class == mot::SemanticClass::kUnknown ||
+        detection.semantic_class == mot::SemanticClass::kUnknown) {
+        return dist;
+      }
+
+      if (track.semantic_class != detection.semantic_class) {
+        return std::nullopt;
+      }
+
+      return dist;
+    } else {
+      return std::nullopt;
+    }
+  }
+
+  template <typename... TrackAlternatives, typename... DetectionAlternatives>
+  auto operator()(
+    const std::variant<TrackAlternatives...> & track,
+    const std::variant<DetectionAlternatives...> & detection) const -> std::optional<float>
+  {
+    return std::visit(
+      [this](const auto & t, const auto & d) { return (*this)(t, d); }, track, detection);
+  }
+
+private:
+  static auto two_dimensional_distance(const mot::CtrvState & lhs, const mot::CtrvState & rhs)
+    -> float
+  {
+    const auto x_diff_sq{
+      std::pow(mot::remove_units(lhs.position_x) - mot::remove_units(rhs.position_x), 2)};
+    const auto y_diff_sq{
+      std::pow(mot::remove_units(lhs.position_y) - mot::remove_units(rhs.position_y), 2)};
+
+    return std::sqrt(x_diff_sq + y_diff_sq);
+  }
+
+  static auto two_dimensional_distance(const mot::CtraState & lhs, const mot::CtraState & rhs)
+    -> float
+  {
+    const auto x_diff_sq{
+      std::pow(mot::remove_units(lhs.position_x) - mot::remove_units(rhs.position_x), 2)};
+    const auto y_diff_sq{
+      std::pow(mot::remove_units(lhs.position_y) - mot::remove_units(rhs.position_y), 2)};
+
+    return std::sqrt(x_diff_sq + y_diff_sq);
+  }
+};
+
+struct MetricSe2
```

Comment:
Let's add small documentation what this class is about

Reply:
Done